library/acl.inc

```diff
@@ -835,4 +835,48 @@
         else
             return false;
     }
+
+  // This generates an HTML options list for all ACOs.
+  // The caller inserts this between <select> and </select> tags.
+  //
+  function gen_aco_html_options($default='') {
+    $s = '';
+    $gacl = new gacl_api();
+    // collect and sort all aco objects
+    $list_aco_objects = $gacl->get_objects(NULL, 0, 'ACO');
+    ksort($list_aco_objects);
+    foreach ($list_aco_objects as $seckey => $dummy) {
+      if (empty($dummy)) continue;
+      asort($list_aco_objects[$seckey]);
+      $aco_section_data = $gacl->get_section_data($seckey, 'ACO');
+      $aco_section_title = $aco_section_data[3];
+      $s .= "<optgroup label='" . xla($aco_section_title) . "'>\n";
+      foreach($list_aco_objects[$seckey] as $acokey) {
+        $aco_id = $gacl->get_object_id($seckey, $acokey,'ACO');
+        $aco_data = $gacl->get_object_data($aco_id, 'ACO');
+        $aco_title = $aco_data[0][3];
+        $optkey = "$seckey|$acokey";
+        $s .= "<option value='" . attr($optkey) . "'";
+        if ($optkey == $default) $s .= ' selected';
+        $s .= ">" . xlt($aco_title) . "</option>";
+      }
+      $s .= "</optgroup>";
+    }
+    return $s;
+  }
+
+  // Permissions check for a specified encounter form type.
+  // Note $return_value may be an array of return values.
+  //
+  function acl_check_form($formdir, $user='', $return_value='') {
+    require_once(dirname(__FILE__) . '/registry.inc');
```

Comment:
Just noting that including scripts within functions can sometimes cause difficult to debug scope issues.